src/store/index.ts

```diff
@@ -205,11 +168,19 @@ const {
 
     addSelectedNode(state, nodeID: string) {
       state.selectedNodes = new Set(state.selectedNodes.add(nodeID));
+
+      if (state.provenance !== null) {
+        updateProvenanceState(state, 'Select Node');
+      }
```

Comment:
This is nice and compact, but something about it feels funny. The VueX state is a reactive object, so you could create something like our multimatrix "event bus" to respond to state changes with its own actions, such as these side-effectful updates of the Trrack state. You could even equip that "bus" with computed props that would provide JSON representations of bits of the VueX state and conceptually streamline the whole thing.

If you decide on that, it can be done in a followup. But since you're testing out an approach here for the first time, I thought I'd mention it.

Reply:
I don't really see how that implementation would be much different from what we already have, but I could just not be understanding what you're saying. It sounds to me like you'd like me to replace this code chunk with a similar code chunk, something like

```
eventBus.updateProvenance()
```

to update the state whenever something changes. If that's the case, I don't think there's much utility. If I'm wrong, would you provide me a bit of pseudo-code so I can see what you mean?

Reply:
Well the idea would be to just let the VueX store manage state, and rely on *other* components to receive updates and perform these side effects.

If you look at any of the components that currently use VueX state, you'll see that you're doing stuff like creating computed props based on state values, then reacting to those locally. What I'm suggesting is to create a controller component that behaves similarly, that would wrap Trrack and watch the VueX values in order to trigger these sorts of actions. Since it's not a "Vue component" in the sense of rendering stuff to the DOM, you could create such a thing almost exactly the way we created that stopgap event bus component.

So, just to summarize, it's not a matter of replacing the code you have with something like `eventBus.updateProvenance()`; it would be closer to supplying the VueX store to `eventBus` when you construct it, then having computed props in that component, and using something like `watch()` or `watchEffect()` in that component to respond to VueX state changes. The major upshot here is reusability: if my design idea actually works, you'll have a fairly general component that can be used in any application that uses both Trrack and VueX in order to add provenance to any Vue app.

What you've got here is simple enough to justify including as-is, but this larger design idea I'm describing would scale better (in case you had other libraries you wanted to bring in in a similar way) and it could be replicated in a modular way to use elsewhere. As I mentioned above, we could roll with this for now and revisit this design in a followup if you feel its sufficiently valuable.

I hope this made sense. I drank two cups of coffee today instead of my usual single cup, and my wife says I am more chatty than usual :upside_down_face:

Reply:
Thanks for the ideas, Roni! It does sound like this would be something worth exploring in a follow up PR. I'm curious how we could initialize the "eventBus" component, since we can't use a traditional vue component, like you said. I guess we could just mount it somewhere like App.vue (the highest level component that mounts everything else) and then pass the VueX store. On the whole, though, it does seem like this would be useful in the case where we have more libraries like this.

Reply:
I think with composition API, you can just declare reactive data anywhere, and it will "just work". I'm not super sure about that though. Jake may have a better understanding.